The following codes were written by Tianye Jia and Sylvane Desrivières, and revised and reviewed by Xinyang Yu and Antoine Weihs. If you have any questions related to these analyses, please contact us at sylvane.desrivieres@kcl.ac.uk or xinyang.1.yu@kcl.ac.uk. 

Files required for these analyses

**Methylation data**

The following data, generated by our ENIGMA-Epigenetics QC protocol will be used: 
"./Quan-norm.rda" (Beta values after quantile normalization)
"./fast_svd.rda" (Principal components of beta values)
"./cellcount.rda" (Estimated cell type proportion)
"./RGset.rda" (Raw dataset after QC)

**Subcortical volumes**

These should be the LandRvolumes.csv file containing the subcortical brain volumes (after quality control) used for ENIGMA2. Please, provide brain volumes extracted with FSL if possible. The LandRvolumes.csv file should look like this:

 ![image](https://github.com/XinyangYu918/DNA-methylation-and-mental-health/assets/52769576/9ead7df9-057d-488c-94b2-0bc1dd0bd16a)


**Covariates file**

The same files used for ENIGMA2 will also be used. This comma-delimited (.csv) text file should contain the following columns: SubjID, Age, Sex. Additional columns for dummy covariates (i.e., a covariate to control for different acquisition sites, if applicable), etc are optional. In addition, the first 4 principal components of the beta value, i.e., from "./fast_svd.rda", and the first two components of estimated cell-type proportion, i.e. from "./cellcount.rda", will also be included as control variables. The relevant code for combining data will be provided.

If your cohort has only healthy controls, only patients and twin study, this file should be named SubCortCovs.csv and look like this:
![image](https://github.com/XinyangYu918/DNA-methylation-and-mental-health/assets/52769576/67166f7c-dd0c-451f-a6d1-ee4b156ea4b2)

For your cohort that has both patients and healthy controls, you should include a covariate called "AffectionStatus", coded as a binary indicator variable where Controls = 0 and Patients = 1. The final file, saved as SubCortCovs.csv, should have the following columns at a minimum: SubjID, Age, Sex, AffectionStatus. Additional columns for dummy covariates (i.e., a covariate to control for different acquisition sites, if applicable) are optional. Note we will have three outputs for the whole sample, the case individuals only and the control individuals only.

*Note 1: Please make sure that missing data has been recorded as NA in these files!! Self-coded missing data should be transformed to NA through Data== -9 <- NA, where Data should be replaced with the name of your data file in question and -9 should be replaced with your self-coded missing data value. 
*Note 2: Sex must be specified as follows: (Males=1, Females=2), and "FID" and "IID" should be named exactly the same in all files.
*Note 3: If population stratification is of reasonable concern, the self-reported ethnicity information should be included in the covariates file, i.e. as dummy variables. As an alternative, the first 4 MDS of genetic data can also be included. 
 

#### The following codes are used to perform QC and run EWAS analysis with subcortical volumes ####
library(minfi)
# Preparing files
# Provide cohort name
cohort = "IMAGEN" #change for the name of your cohort

# Probe quality check
load("./Quan-norm.rda")
load("./RGset.rda")

# Add SNP info to the data 
objectWithSNPinfo <- addSnpInfo(object)

# Drop probes that contain either an SNP at the CpG interrogation or at the single nucleotide extension
objectSNPQCed <- dropLociWithSnps(objectWithSNPinfo, snps=c("SBE", "CpG", "Probe"), maf=0.05)

detP <- detectionP(RGset)
Match1 <- match(colnames(objectSNPQCed),colnames(detP))
Match2 <- match(rownames(objectSNPQCed),rownames(detP))
detPSNPQCed <- detP[Match2[!is.na(Match2)],Match1[!is.na(Match1)]]
rm(detP); rm(Match1); rm(Match2);

failed <- detPSNPQCed >0.01
beta <- getBeta(objectSNPQCed)

# Drop probes that failed quality control via the detection p-value in greater than 20% of samples
failedCG02 <- rowMeans(failed)>0.2

# Get the list of non-variable CpG sites i.e. those where beta values for all samples are ≤20% or ≥80%
ProbeInvar <- (rowSums(beta<=0.2)==ncol(beta))|(rowSums(beta>=0.8)==ncol(beta)) 
ListInvarProbe <- rownames(beta)[which(ProbeInvar)] 

# Remove sex chromosome probes
keepIndex=!seqnames(objectSNPQCed)%in%c("chrX","chrY")
rm(objectSNPQCed)
keepIndex <- keepIndex&(!failedCG02)
rm(failedCG02)
beta[failed] <- NA
rm(failed)
betaQC <- beta[which(keepIndex),]
rm(beta, keepIndex)

# Reformat beta value to match the format for EWAS analysis
Methy <- as.data.frame(t(betaQC))
Methy$Subject <- colnames(betaQC)
rm(betaQC)
Methy <- Methy[,c(ncol(Methy),1:(ncol(Methy)-1))]
MethyName <- colnames(Methy)[-c(1)]

# Load and format covariates
load("./fast_svd.rda")
PC_Beta <- as.data.frame(ss$v[,1:4])
PC_Beta$Subject <- Methy$Subject

load("./cellcount.rda")
tmp <- prcomp(cellcount)
pc_cell <- as.data.frame(tmp$x[,1:2])
pc_cell$Subject <- rownames(pc_cell)

# Load and format Subcortical volumes
Raw_Struc <- read.csv("./LandRvolumes.csv", header=T) 
Struc <- matrix(data=0, ncol=8, nrow=nrow(Raw_Struc)) 
colnames(Struc) <- c("Subject","Mthal", "Mcaud", "Mput", "Mpal", "Mhippo", "Mamyg", "Maccumb")
Struc <- as.data.frame(Struc) 
Struc$Subject = Raw_Struc$SubjID 
Struc$Mthal <- rowMeans(Raw_Struc[,c("Lthal","Rthal")])
Struc$Mcaud <- rowMeans(Raw_Struc[,c("Lcaud","Rcaud")])
Struc$Mput <- rowMeans(Raw_Struc[,c("Lput","Rput")])
Struc$Mpal <- rowMeans(Raw_Struc[,c("Lpal","Rpal")])
Struc$Mhippo <- rowMeans(Raw_Struc[,c("Lhippo","Rhippo")])
Struc$Mamyg <- rowMeans(Raw_Struc[,c("Lamyg","Ramyg")])
Struc$Maccumb <- rowMeans(Raw_Struc[,c("Laccumb","Raccumb")])
StrucName <- colnames(Struc)[-1]
ICV <- Raw_Struc[,c(1,ncol(Raw_Struc))] 
colnames(ICV)[1] <- "Subject" 
rm(Raw_Struc)

# Read and combine Covariates Data
Raw_Cov <- read.csv("./SubCortCovs.csv", header=T) 
Raw_Cov<-Raw_Cov[,-c(1)] 
Raw_Cov$Age_Square <- (Raw_Cov$Age)^2 
colnames(Raw_Cov)[1] <- "Subject"

Cov <- merge(Raw_Cov, PC_Beta, by="Subject", all=F) 
Cov <- merge(Cov, pc_cell, by="Subject", all=F) 
Cov <- merge(Cov, ICV, by="Subject", all=F) 
CovName <- colnames(Cov)[-c(1)] 

# Save age information for further analysis
AgeInformation <- data.frame(min(Cov$Age),max(Cov$Age),mean(Cov$Age))
names(AgeInformation) <- c("MinAge", "MaxAge", "MeanAge")

# Merge files for EWAS, make sure the inputs are in the same order for participants
Data <- merge(Cov, Struc, by="Subject", all=F)
Match <- match(Methy$Subject, Data$Subject)

Cov <- Data[Match[!is.na(Match)],2:(length(CovName)+1)]
Struc <- Data[Match[!is.na(Match)],-c(1:(length(CovName)+1))]
Methy <- Methy[!is.na(Match),-c(1)]

# EWAS analysis: linear regression between DNA methylation and subcortical volumes
Num_Methy <- ncol(Methy) 
Num_Cov <- ncol(Cov)
Num_Struc <- ncol(Struc)

# Prepare the output files 
Origin_Beta <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_Beta) <- colnames(Struc)
rownames(Origin_Beta) <- colnames(Methy)
Origin_SE <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_SE) <- colnames(Struc)
rownames(Origin_SE) <- colnames(Methy)
Origin_P <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_P) <- colnames(Struc)
rownames(Origin_P) <- colnames(Methy)
Origin_N <- matrix(data=NA,nrow=Num_Methy,ncol=Num_Struc,byrow=F,dimnames=NULL)
colnames(Origin_N) <- colnames(Struc)
rownames(Origin_N) <- colnames(Methy)

# The following codes should be run by all cohorts
Covar <- matrix(unlist(Cov), ncol=ncol(Cov), byrow=F)
for (i in 1:Num_Methy) {
  for (j in 1:Num_Struc ) {
    Out <- summary(lm(Methy[,i] ~ Covar + Struc[,j]))
    Origin_Beta[i,j] <- Out$coefficients[nrow(Out$coefficients),1]
    Origin_SE[i,j] <- Out$coefficients[nrow(Out$coefficients),2]
    Origin_P[i,j] <- Out$coefficients[nrow(Out$coefficients),4]
    Origin_N[i,j] <- Out$df[1]+Out$df[2]
  }
}
# Save model description
model <- paste0("Methy[,i] ~ ",paste0(names(Cov),collapse = "+"),"+Struc [,j]")

# Save output file
save(Origin_Beta, Origin_SE, Origin_P,model,Origin_N,ListInvarProbe, AgeInformation, file = paste("./Output_of_",cohort, "_Methylation_and_All_Subcortical_Structure.RData",sep=""), compress = T)


# The following codes should be run by case-control cohorts
# Prepare the output files 
Origin_Beta <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_Beta) <- colnames(Struc)
rownames(Origin_Beta) <- colnames(Methy)
Origin_SE <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_SE) <- colnames(Struc)
rownames(Origin_SE) <- colnames(Methy)
Origin_P <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_P) <- colnames(Struc)
rownames(Origin_P) <- colnames(Methy)
Origin_N <- matrix(data=NA,nrow=Num_Methy,ncol=Num_Struc,byrow=F,dimnames=NULL)
colnames(Origin_N) <- colnames(Struc)
rownames(Origin_N) <- colnames(Methy)

Affect <- colnames(Cov)=="AffectionStatus"

# Save age information for cases and controls separately
AgeInformation$CaseMinAge <- min(Cov$Age[Cov$AffectionStatus == 1])
AgeInformation$CaseMaxAge <- max(Cov$Age[Cov$AffectionStatus == 1])
AgeInformation$CaseMeanAge <- mean(Cov$Age[Cov$AffectionStatus == 1])
AgeInformation$ControlMinAge <- min(Cov$Age[Cov$AffectionStatus == 0])
AgeInformation$ControlMaxAge <- max(Cov$Age[Cov$AffectionStatus == 0])
AgeInformation$ControlMeanAge <- mean(Cov$Age[Cov$AffectionStatus == 0])

if (sum(Affect)==1) 
{
Status <- c("Case", "Control")
for (k in 1:2) {
Ind <- Covar[,Affect]==(2-k) 
for (i in 1:Num_Methy) {
for (j in 1:Num_Struc) {
Out <- summary(lm(Methy[Ind,i]~Covar[Ind,!Affect]+Struc[Ind,j]))         
Origin_Beta[i,j] <- Out$coefficients[nrow(Out$coefficients),1]
Origin_SE[i,j] <- Out$coefficients[nrow(Out$coefficients),2]
Origin_P[i,j] <- Out$coefficients[nrow(Out$coefficients),4]
Origin_N[i,j] <- Out$df[1]+Out$df[2]
}
}

# Save model description
model <- paste0("Methy[Ind,i] ~ ",paste0(names(Cov)[!Affect],collapse = "+"),"+ Struc [Ind,j]")

# Save output file
save(Origin_Beta, Origin_SE, Origin_P, Origin_N, model,ListInvarProbe, AgeInformation, file = paste("./Output_of_",cohort,"_",Status[k], "Individuals_Methylation_and_All_Subcortical_Structure.RData", sep=""), compress = T)
    }
}
