The following codes were written by Tianye Jia and Sylvane Desrivières, and revised and reviewed by Xinyang Yu and Antoine Weihs. If you have any questions related to these analyses, please contact us at sylvane.desrivieres@kcl.ac.uk or xinyang.1.yu@kcl.ac.uk. 

Files required for these analyses

Methylation data – The following data, generated by our ENIGMA-Epigenetics QC protocol will be used: 
"./Quan-norm.rda" (Beta values after quantile normalization)
"./fast_svd.rda" (Principal components of beta values)
"./cellcount.rda" (Estimated cell type proportion)
"./RGset.rda" (Raw dataset after QC)

Subcortical volumes ¬– These should be the LandRvolumes.csv file containing the subcortical brain volumes (after quality control) used for ENIGMA2. Please, provide brain volumes extracted with FSL if possible. The LandRvolumes.csv file should look like this:

 

Covariates file–The same files used for ENIGMA2 will also be used. This comma delimited (.csv) text file should contain the following columns: SubjID, Age, Sex. Additional columns for dummy covariates (i.e., a covariate to control for different acquisitions sites, if applicable), etc is optional. In addition, the first 4 principal components of the beta value, i.e., from "./fast_svd.rda", and the first two components of estimated cell-type proportion, i.e. from "./cellcount.rda", will also be included as control variables. The relevant code of combining data will be provided.
•	If your cohort has only healthy controls, only patients and twin study, this file should be named SubCortCovs.csv and look like this:
 
•	For your cohort has both patients and healthy controls, you should include a covariate called "AffectionStatus", coded as a binary indicator variable where Controls = 0 and Patients = 1. The final file, saved as SubCortCovs.csv, should have the following columns at a minimum: SubjID, Age, Sex, AffectionStatus. Additional columns for dummy covariates (i.e., a covariate to control for different acquisitions sites, if applicable) is optional. Note we will have three outputs for the whole sample, the case individuals only and the control individuals only.
•	Note 1: !! Please make sure that missing data has been recoded as NA in these files!! Self-coded missing data should be transformed to NA through Data== -9 <- NA, where Data should be replaced with the name of your data file in question and -9 should be replaced with your self-coded missing data value. 
•	Note 2: Sex must be specified as follows: (Males=1, Females=2), and "FID" and "IID" should be named exactly the same in all files.
•	Note 3: If population stratification is of reasonable concern, the self-reported ethnicity information should be included in the covariates file, i.e. as dummy variables. As an alternative, the first 4 MDS of genetic data can also be included. 
 

##################
#####Preparing files#
##################
#Provide cohort name
cohort = "IMAGEN" #change for the name of your cohort


###Probe Quality Check####
library(minfi)

load("./Quan-norm.rda") #variable object will be loaded
load("./RGset.rda") #variable RGset will be loaded
#Adding SNP info to the data 
objectWithSNPinfo <- addSnpInfo(object)

#Dropping probes that contain either a SNP at the CpG interrogation or at the single nucleotide extension
objectSNPQCed<- dropLociWithSnps(objectWithSNPinfo, snps=c("SBE","CpG", "Probe"), maf=0.05)

detP <- detectionP(RGset)
Match1 <- match(colnames(objectSNPQCed),colnames(detP))
Match2 <- match(rownames(objectSNPQCed),rownames(detP))
detPSNPQCed <- detP[Match2[!is.na(Match2)],Match1[!is.na(Match1)]]
rm(detP); rm(Match1); rm(Match2);

failed <- detPSNPQCed >0.01
beta <- getBeta(objectSNPQCed)
###Drop probes that failed quality control via the detection p-value in greater than 20% of samples
failedCG02 <- rowMeans(failed)>0.2

####Get the list non-variable CpG sites i.e. those where beta values for all samples are ≤20% or ≥80%
ProbeInvar <- (rowSums(beta<=0.2)==ncol(beta))|(rowSums(beta>=0.8)==ncol(beta)) #Mark probes with either all beta value <=0.2 or all beta value>=0.80
ListInvarProbe <- rownames(beta)[which(ProbeInvar)] #Generate a list of probes marked as invariant.
#This list will be included in the output file that you will send us
rm(ProbeInvar);

###Remove sex chromosome probes#####
keepIndex=!seqnames(objectSNPQCed)%in%c("chrX","chrY") #mark probes on X and Y chromosome
rm(objectSNPQCed);
keepIndex <- keepIndex&(!failedCG02) #Combine with failed probes
rm(failedCG02);
beta[failed] <- NA #Remove all probes with detected P-value >0.01.
rm(failed);
betaQC <- beta[which(keepIndex),] #Remove probes
rm(beta); rm(keepIndex);
##Reformat beta value to match the format suggested above
Methy <- as.data.frame(t(betaQC)) #load quantile normalized methylation data
Methy$Subject <- colnames(betaQC) #add subject ID to the methylation data as the last column
rm(betaQC);
Methy <- Methy[,c(ncol(Methy),1:(ncol(Methy)-1))] #re-order the column and making the subject ID as the first column
MethyName <- colnames(Methy)[-c(1)] #output probe names

load("./fast_svd.rda") #load principal components of beta value
PC_Beta <- as.data.frame(ss$v[,1:4]) #generate a variable with first four principle components of methylation
PC_Beta$Subject <- Methy$Subject #add subject ID to the component variable 
load("./cellcount.rda") #load estimated cell type proportion
tmp <- prcomp(cellcount) #generate principle components of cell count
pc_cell <- as.data.frame(tmp$x[,1:2]) #Generate a variable with subject ID and first 2 PCs of cellcount
pc_cell$Subject <- rownames(pc_cell) #Add row names

###Read and format Subcortical volumes ####
Raw_Struc <- read.csv("./LandRvolumes.csv", header=T) #Please ensure the format of the  ‘LandRvolumes.csv’ file is as illustrated above. 
Struc <- matrix(data=0, ncol=8, nrow=nrow(Raw_Struc)) #Generate a new matrix for the mean of L/R subcortical brain regions
colnames(Struc) <- c("Subject","Mthal", "Mcaud", "Mput", "Mpal", "Mhippo", "Mamyg", "Maccumb") #Provide column names for the mean subcortical brain volumes
Struc <- as.data.frame(Struc) #set data as data.frame
Struc$Subject = Raw_Struc$SubjID #assign subject ID 
Struc$Mthal <- rowMeans(Raw_Struc[,c("Lthal","Rthal")]); #calculate mean Thalamus volume
Struc$Mcaud <- rowMeans(Raw_Struc[,c("Lcaud","Rcaud")]); #calculate mean Caudate volume
Struc$Mput <- rowMeans(Raw_Struc[,c("Lput","Rput")]); #calculate mean Putamen volume
Struc$Mpal <- rowMeans(Raw_Struc[,c("Lpal","Rpal")]); #calculate mean Pallidum volume
Struc$Mhippo <- rowMeans(Raw_Struc[,c("Lhippo","Rhippo")]); #calculate mean volume Hippocampus 
Struc$Mamyg <- rowMeans(Raw_Struc[,c("Lamyg","Ramyg")]); #calculate mean Amygdala volume
Struc$Maccumb <- rowMeans(Raw_Struc[,c("Laccumb","Raccumb")]); #calculate mean Nucleus Accumbens volume
StrucName <- colnames(Struc)[-1] #column names of subcortical brain volumes
ICV <- Raw_Struc[,c(1,ncol(Raw_Struc))] #generate a separate variable for ICV
colnames(ICV)[1] <- "Subject" #rename the column for easier merging
rm(Raw_Struc);

###Read and Combine Covariates Data#####
Raw_Cov <- read.csv("./SubCortCovs.csv", header=T) #load covariates file as illustrated above, please remember to include the extra column of ‘AffectStatus’ for case/control sample. 
Raw_Cov<-Raw_Cov[,-c(1)] #Assume it is the IID that is equal to the Subject ID from Structure and Methylation data
Raw_Cov$Age_Square <- (Raw_Cov$Age)^2 #Add age^2 as the last column of Raw_Cov 
colnames(Raw_Cov)[1] <- "Subject"
Cov <- merge(Raw_Cov, PC_Beta, by="Subject", all=F) #combine covariates and PCs of methylation data
Cov <- merge(Cov, pc_cell, by="Subject", all=F) #combine covariates and PCs of cell count
Cov <- merge(Cov, ICV, by="Subject", all=F) #combine covariates and ICV
CovName <- colnames(Cov)[-c(1)] #names of covariates
rm(PC_Beta); rm(pc_cell); rm(ICV); rm(Raw_Cov);

###Generate a dataframe to save age information for further analysis#####
AgeInformation <- data.frame(min(Cov$Age),max(Cov$Age),mean(Cov$Age))
names(AgeInformation) <- c("MinAge","MaxAge","MeanAge")

###Generating Files with Complete Data across All Data#####
Data <- merge(Cov, Struc, by="Subject", all=F) #combine covariates and structure data
Match <- match(Methy$Subject, Data$Subject) #match methylation sample with combined covariates and structure sample

##Generate new covariates, structure and methylation data, which have had their individuals matched
Cov <- Data[Match[!is.na(Match)],2:(length(CovName)+1)]  #generate covariates with matched individual
Struc <- Data[Match[!is.na(Match)],-c(1:(length(CovName)+1))] #generate structure data with matched individual
Methy <- Methy[!is.na(Match),-c(1)] #generate methylation data with matched individual
rm(Data); gc()

#####Linear regression Analysis######
Num_Methy <- ncol(Methy) 
Num_Cov <- ncol(Cov)
Num_Struc <- ncol(Struc)

##Preparing the output files 
Origin_Beta <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_Beta) <- colnames(Struc)
rownames(Origin_Beta) <- colnames(Methy)
Origin_SE <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_SE) <- colnames(Struc)
rownames(Origin_SE) <- colnames(Methy)
Origin_P <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_P) <- colnames(Struc)
rownames(Origin_P) <- colnames(Methy)
Origin_N <- matrix(data=NA,nrow=Num_Methy,ncol=Num_Struc,byrow=F,dimnames=NULL)
colnames(Origin_N) <- colnames(Struc)
rownames(Origin_N) <- colnames(Methy)

##################################################################
##Association analysis: the following section should be run by ALL cohorts 
##################################################################

Covar <- matrix(unlist(Cov), ncol=ncol(Cov), byrow=F)
for (i in 1:Num_Methy) {
  for (j in 1:Num_Struc ) {
    Out <- summary(lm(Methy[,i] ~ Covar + Struc[,j]))
    Origin_Beta[i,j] <- Out$coefficients[nrow(Out$coefficients),1]
    Origin_SE[i,j] <- Out$coefficients[nrow(Out$coefficients),2]
    Origin_P[i,j] <- Out$coefficients[nrow(Out$coefficients),4]
    Origin_N[i,j] <- Out$df[1]+Out$df[2]
  }
}
## Save model description
model <- paste0("Methy[,i] ~ ",paste0(names(Cov),collapse = "+"),"+Struc [,j]")

###########################################################################
#######################Parrallel Alternative (only Linux)##########################
###########################################################################

doOne <- function(i, j)
{
	tryCatch({
		Out <- summary(lm(Methy[,i]~Covar+Struc[,j]))
		return(as.data.frame(t(c(Out$coefficients[nrow(Out$coefficients),1], Out$coefficients[nrow(Out$coefficients),2], Out$coefficients[nrow(Out$coefficients),4], Out$df[1]+Out$df[2], names(Methy)[i]))))
	}, error = function(e) {
    cat("Warning: LM failed for probe ", names(Methy)[i], " and subcortical region ", names(Struc)[j], "\n")
    print(e)
    return(t(c(NA, NA, NA, NA, NA)))
	}, finally={})
}

for (j in 1:Num_Struc)
{
	print(paste0("Running brain region (", j, "/", Num_Struc, "): ", names(Struc)[j]))
	res = pbmclapply(1:Num_Methy, doOne, j=j, mc.cores = 15)
	res = dplyr::bind_rows(res)
	rownames(res) = res[,5]
	res = res[colnames(Methy),]
	Origin_Beta[,i] = res[,1]
	Origin_SE[,i] = res[,2]
	Origin_P[,i] = res[,3]
	Origin_N[,i] = res[,4]
}

###########################################################################
###########################################################################


###Save output file
save(Origin_Beta, Origin_SE, Origin_P,model,Origin_N,ListInvarProbe, AgeInformation, file=paste("./Output_of_",cohort, "_Methylation_and_All_Subcortical_Structure.RData",sep=""),compress=T)
#This file should contain output for all subcortical volumes, including the beta, SE, and P-values for all 7 subcortical brain volumes, a list of invariant probes, as well as age information about the cohort.

rm(Origin_Beta, Origin_SE, Origin_P,model,Origin_N); gc()

###########################################################################
##Cohorts with patients should also run the following section, in addition to the above ##
###########################################################################

Origin_Beta <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_Beta) <- colnames(Struc)
rownames(Origin_Beta) <- colnames(Methy)
Origin_SE <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_SE) <- colnames(Struc)
rownames(Origin_SE) <- colnames(Methy)
Origin_P <- matrix(data= NA, nrow=Num_Methy, ncol= Num_Struc, byrow=F, dimnames=NULL)
colnames(Origin_P) <- colnames(Struc)
rownames(Origin_P) <- colnames(Methy)
Origin_N <- matrix(data=NA,nrow=Num_Methy,ncol=Num_Struc,byrow=F,dimnames=NULL)
colnames(Origin_N) <- colnames(Struc)
rownames(Origin_N) <- colnames(Methy)

##The script below will perform associations for cases and controls separately.
Affect <- colnames(Cov)=="AffectionStatus"

###Get and save the age information for cases and controls separately, please make sure affection status are coded as 0 (control) and 1 (case)####
AgeInformation$CaseMinAge <- min(Cov$Age[Cov$AffectionStatus == 1])
AgeInformation$CaseMaxAge <- max(Cov$Age[Cov$AffectionStatus == 1])
AgeInformation$CaseMeanAge <- mean(Cov$Age[Cov$AffectionStatus == 1])
AgeInformation$ControlMinAge <- min(Cov$Age[Cov$AffectionStatus == 0])
AgeInformation$ControlMaxAge <- max(Cov$Age[Cov$AffectionStatus == 0])
AgeInformation$ControlMeanAge <- mean(Cov$Age[Cov$AffectionStatus == 0])

if (sum(Affect)==1) { #Check if there is a column named "AffectionStatus", i.e. a case-control study 
    Status <- c("Case","Control")
    for (k in 1:2) {
        Ind <- Covar[,Affect]==(2-k) #AffectionStatus should be coded as 1 (Case) and 0 (Control)
for (i in 1:Num_Methy) {
            for (j in 1:Num_Struc) {
            Out <- summary(lm(Methy[Ind,i]~Covar[Ind,!Affect]+Struc[Ind,j]))         
            Origin_Beta[i,j] <- Out$coefficients[nrow(Out$coefficients),1]
            Origin_SE[i,j] <- Out$coefficients[nrow(Out$coefficients),2]
            Origin_P[i,j] <- Out$coefficients[nrow(Out$coefficients),4]
            Origin_N[i,j] <- Out$df[1]+Out$df[2]
            }
        }
## Save model description
model <- paste0("Methy[Ind,i] ~ ",paste0(names(Cov)[!Affect],collapse = "+"),"+ Struc [Ind,j]")
        ###Save output file
        save(Origin_Beta, Origin_SE, Origin_P, Origin_N,model,ListInvarProbe, AgeInformation, file=paste("./Output_of_",cohort,"_",Status[k],"Individuals_Methylation_and_All_Subcortical_Structure.RData",sep=""),compress=T)
#This file should contain output for all subcortical volumes, including the beta, SE, and P-values for all 7 subcortical brain volumes, a list of invariant probes, as well as age information for both cases and controls.
    }
}

#For case control studies, all three RData files (i.e., corresponding to outputs from the full sample, cases only and controls only) for the subcortical structure should be sent to us.  For all other cohorts, one RData file, corresponding to outputs from the full sample, should be sent to us.

#Other Information Required: Apart from the relevant output files, we will also need information of sample size and covariates included.

